native/membrane_videocompositor/src/elixir_bridge/elixir_structs.rs

```diff
@@ -136,3 +140,328 @@ impl From<ElixirCropping> for Box<dyn TextureTransformation> {
+                                })?,
+                                nodes,
+                                objects,
+                            )
+                            .map(|node| (internal_name.clone(), node))
+                        })
+                        .collect::<Result<HashMap<_, _>, _>>()?;
+
+                    let result = Arc::new(Node::Layout {
+                        resolution: resolution.clone(),
+                        implementation: *implementation,
+                        inputs,
+                    });
+                    nodes.insert(name.clone(), result.clone());
+                    Ok(result)
+                }
+            }
+        }
+
+        parse_object(&self.output, &final_object, &mut nodes, &objects)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SceneParsingError {
+    #[error("cycle detected in the scene graph")]
+    CycleDetected,
+
+    #[error("the scene graph contains an object name that is not defined anywhere: {0:?}")]
+    UndefinedName(ObjectName),
+
+    #[error("object name {0:?} is defined multiple times in the scene graph")]
+    DuplicateNames(ObjectName),
+
+    #[error("the scene graph contains two InputVideos referencing the same Membrane.Pad: {0}")]
+    DuplicatePadReferences(PadRef),
+
+    #[error("the scene graph contains an object which is not used in compositing: {0:?}")]
+    UnusedObject(ObjectName),
+}
+
+impl rustler::Encoder for SceneParsingError {
+    fn encode<'a>(&self, env: rustler::Env<'a>) -> rustler::Term<'a> {
+        match self {
+            SceneParsingError::CycleDetected => rustler::Atom::from_str(env, "cycle_detected")
+                .unwrap()
+                .encode(env),
+
+            SceneParsingError::UndefinedName(name) => (
+                rustler::Atom::from_str(env, "undefined_name").unwrap(),
+                name,
+            )
+                .encode(env),
+
+            SceneParsingError::DuplicateNames(name) => (
+                rustler::Atom::from_str(env, "duplicate_names").unwrap(),
+                name,
+            )
+                .encode(env),
+
+            SceneParsingError::DuplicatePadReferences(pad) => (
+                rustler::Atom::from_str(env, "duplicate_pad_refs").unwrap(),
+                pad,
+            )
+                .encode(env),
+
+            SceneParsingError::UnusedObject(name) => {
+                (rustler::Atom::from_str(env, "unused_name").unwrap(), name).encode(env)
+            }
+        }
+    }
+}
+
+impl From<SceneParsingError> for rustler::Error {
+    fn from(err: SceneParsingError) -> Self {
+        Self::Term(Box::new(err))
+    }
+}
+
+impl TryInto<crate::scene::Scene> for Scene {
+    type Error = SceneParsingError;
+
+    fn try_into(self) -> Result<crate::scene::Scene, Self::Error> {
+        self.validate()?;
+
+        let node = self.convert_to_graph()?;
+
+        Ok(crate::scene::Scene { final_node: node })
+    }
+}
```

Comment:
Shouldn't this logic be in `scene_validation.rs`? If it is possible to have multiple modules with `rustler` translation, I would move that there.